src/Symfony/Component/Serializer/Attribute/MaxDepth.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Attribute;
+
+/**
+ * Defines the maximum serialization depth for the property.
+ *
+ * When the maximum depth is reached, the $maxDepthReachedFormatter callable is called if it has been defined.
+ *
+ * The first argument of that callable must be the input data.
+ * Then, it is possible to inject the {@see Symfony\Component\Serializer\Serialize\Config\SerializeConfig}
+ * and services thanks to their FQCN.
+ *
+ * It must return the new data.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental in 7.0
+ */
+#[\Attribute(\Attribute::TARGET_PROPERTY)]
+final readonly class MaxDepth
+{
+    /**
+     * @param positive-int $maxDepth
+     * @param ?callable    $maxDepthReachedFormatter
+     */
+    public function __construct(
+        public int $maxDepth,
+        public mixed $maxDepthReachedFormatter = null,
```

Comment:
```suggestion
        public ?callable $maxDepthReachedFormatter = null,
```

Reply:
Same here, properties are unfortunately not allowed to have the `callable` type in PHP.

Reply:
The solution here is to stop using promoted properties (which are just a syntactic sugar for the simple case, not a mandatory feature), to be able to use `?callable` as the constructor argument type and `?Closure` as the property type.


However, I would recommend reconsidering this design: defining a callable in an attribute is quite hard as most ways to define a callable are forbidden in the context of an attribute. And when caching of metadata is involved, it is easier if the metadata is serializable, which is not the case for callables. The formatter callable should probably be provided from the configuration, not from the mapping attribute.

Reply:
Updated to use the `Closure` type.

I agree with your recommendation, but I'm struggling with it.

Indeed, the fact that the encoding/decoding is taking place in two steps ("cache building" step and "running" step), and that these two steps can occur in two separate PHP processes, I cannot be sure that something provided during the "cache building" step will be the same as what is provided during the "running" step. Therefore, I cannot make sure that the formatter will have the same signature for both steps. And if the signature changes between the two steps, it'll break the encoding/decoding.

Another approach could be to allow "formatter" properties to be only strings that are valid public static functions, WDYT?

(Btw, I noticed that the same it is being done here: https://github.com/symfony/symfony/pull/51741/files#diff-2dd48ed7b53254be6400dd49570512b2ca0ebb1808fa010b1c5bfdf543a65915R35)